libs/jwt-tools/src/Jwt/Tools.hs

```diff
@@ -0,0 +1,188 @@
+import Foreign.Ptr (Ptr, nullPtr)
+import Foreign.Storable (peek)
+import Imports
+import Network.HTTP.Types (StdMethod (..))
+import Numeric (readHex)
+import Wire.API.MLS.Credential
+import Wire.API.MLS.Epoch (Epoch (..))
+import Wire.API.User.Client.DPoPAccessToken
+
+data JwtResponse
+
+foreign import ccall unsafe "generate_dpop_access_token"
+  generate_dpop_access_token ::
+    CString ->
+    CString ->
+    Word16 ->
+    CString ->
+    CString ->
+    CString ->
+    CString ->
+    Word16 ->
+    Word64 ->
+    Word64 ->
+    CString ->
+    IO (Ptr JwtResponse)
+
+foreign import ccall unsafe "free_dpop_access_token" free_dpop_access_token :: Ptr JwtResponse -> IO ()
+
+foreign import ccall unsafe "get_error" get_error :: Ptr JwtResponse -> Ptr CUChar
+
+foreign import ccall unsafe "get_token" get_token :: Ptr JwtResponse -> CString
+
+createToken ::
+  CString ->
+  CString ->
+  Word16 ->
+  CString ->
+  CString ->
+  CString ->
+  CString ->
+  Word16 ->
+  Word64 ->
+  Word64 ->
+  CString ->
+  IO (Maybe (Ptr JwtResponse))
+createToken dpopProof user client domain nonce uri method maxSkewSecs expiration now backendKeys = do
+  ptr <- generate_dpop_access_token dpopProof user client domain nonce uri method maxSkewSecs expiration now backendKeys
+  if ptr /= nullPtr
+    then pure $ Just ptr
+    else pure Nothing
+
+getError :: Ptr JwtResponse -> IO (Maybe Word8)
+getError ptr = do
+  let errorPtr = get_error ptr
+  if errorPtr /= nullPtr
+    then Just . fromIntegral <$> peek errorPtr
+    else pure Nothing
+
+getToken :: Ptr JwtResponse -> IO (Maybe String)
+getToken ptr = do
+  let tokenPtr = get_token ptr
+  if tokenPtr /= nullPtr
+    then Just <$> peekCString tokenPtr
+    else pure Nothing
+
+generateDpopToken ::
+  (MonadIO m) =>
+  Proof ->
+  ClientIdentity ->
+  Nonce ->
+  HttpsUrl ->
+  StdMethod ->
+  Word16 ->
+  Epoch ->
+  Epoch ->
+  PEMKeys ->
+  ExceptT DPoPTokenGenerationError m DPoPAccessToken
+generateDpopToken dpopProof cid nonce uri method maxSkewSecs maxExpiration now backendPubkeyBundle = do
+  dpopProofCStr <- toCStr dpopProof
+  uidCStr <- toCStr $ ciUser cid
+  cidCUShort <- case readHex @Word16 (cs $ client $ ciClient cid) of
+    [(a, "")] -> pure a
+    _ -> throwE ClientIdSyntaxError
+  domainCStr <- toCStr $ ciDomain cid
+  nonceCStr <- toCStr nonce
+  uriCStr <- toCStr uri
+  methodCStr <- liftIO $ newCString $ cs $ methodToBS method
+  backendPubkeyBundleCStr <- toCStr backendPubkeyBundle
+
+  let before =
+        createToken
+          dpopProofCStr
+          uidCStr
+          cidCUShort
+          domainCStr
+          nonceCStr
+          uriCStr
+          methodCStr
+          maxSkewSecs
+          (epochNumber maxExpiration)
+          (epochNumber now)
+          backendPubkeyBundleCStr
+
+  let mkAccessToken response = do
+        case response of
+          Nothing -> pure $ Left FfiError
+          Just r -> do
+            mErr <- getError r
+            print mErr
```

Comment:
Leftover? :)

Reply:
:100:

Reply:
nice because we can use applicative composition now: `toResult <$> getError r <*> getToken r`